activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class InsertAll
+    def initialize(model, inserts, **options)
+      raise ArgumentError, "inserts should be an array" unless inserts.is_a?(Array)
+      @model, @connection, @options = model, model.connection, options
+      @keys, @values_list = connection.prepare_keys_and_values_for_insert(inserts, model.table_name)
+    end
+
+    def execute
+      if values_list.empty?
+        ActiveRecord::Result.new([], [])
+      else
+        connection.exec_query to_sql, "Bulk Insert"
+      end
+    end
+
+    def to_sql
+      sql_builder.build_sql(
+        model,
+        keys,
+        values_list,
+        returning: returning,
+        on_duplicate: on_duplicate,
+        unique_by: unique_by,
+        updatable_columns: updatable_columns)
+    end
+
+    private
+      attr_reader :model, :connection, :options, :keys, :values_list
+
+      def sql_builder
+        connection.insert_all_sql_builder
+      end
+
+      def returning
+        returning = options.fetch(:returning) { connection.supports_insert_returning? && primary_keys }
+        returning = false if returning == []
+
+        if returning && !connection.supports_insert_returning?
+          raise ArgumentError, "#{connection.class} does not support :returning"
+        end
+
+        returning
+      end
+
+      def on_duplicate
+        on_duplicate = options.fetch(:on_duplicate, :raise)
+        on_duplicate = :skip if on_duplicate == :update && updatable_columns.empty?
+
+        unless %i{ raise skip update }.member?(on_duplicate)
+          raise NotImplementedError, "#{on_duplicate.inspect} is an unknown value for :on_duplicate. Valid values are :raise, :skip, and :update"
+        end
+
+        if on_duplicate == :skip && !connection.supports_insert_on_duplicate_skip?
+          raise ArgumentError, "#{connection.class} does not support skipping duplicates"
+        end
+
+        if on_duplicate == :update && !connection.supports_insert_on_duplicate_update?
+          raise ArgumentError, "#{connection.class} does not support upsert"
+        end
+
+        on_duplicate
+      end
+
+      def unique_by
+        unique_by = options[:unique_by]
+
+        if unique_by && !connection.supports_insert_conflict_target?
+          raise ArgumentError, "#{connection.class} does not support :unique_by"
+        end
+
+        unique_by
+      end
+
+      def updatable_columns
+        @updatable_columns ||= keys - readonly_columns - unique_by_columns
```

Comment:
I don't think we need to memoize here?

Reply:
This method gets called twice — same for `conflict_columns` — those are the two places where I still use memoization in this version. Both methods do 6 or so operations and get called twice; but it's not like either is doing I/O. Do you think I should axe the memoization?

Reply:
Yeah, that definitely does not seem worth the memoization to me.

Reply:
👍 I'll change that and push it up on a minute.

Reply:
✅ unmemoized